Source/WebCore/html/ImageDataArray.cpp

```diff
@@ -29,6 +29,185 @@
+{
+    return tryCreate(length, storageFormat, true);
+}
+
+std::optional<ImageDataArray> ImageDataArray::tryCreateFromBufferOrZero(size_t length, ImageDataStorageFormat storageFormat, const uint8_t* buffer, size_t bytes)
+{
+    std::optional<ImageDataArray> array = tryCreate(length, storageFormat, false);
+    if (array) {
+        visitArrayBufferView(
+            array->arrayBufferView(),
+            [&](auto& array) {
+                memcpy(array.data(), buffer, bytes);
+            });
+    }
+    return array;
+}
+
+constexpr bool ImageDataArray::isSupported(JSC::TypedArrayType typedArrayType)
+{
+    switch (typedArrayType) {
+    case JSC::TypeUint8Clamped:
+    case JSC::TypeFloat16:
+        return true;
+    default:
+        return false;
+    }
+}
+
+constexpr bool ImageDataArray::isSupported(const JSC::ArrayBufferView& arrayBufferView)
+{
+    return isSupported(arrayBufferView.getType());
+}
+
+ImageDataStorageFormat ImageDataArray::storageFormat() const
+{
+    switch (m_arrayBufferView->getType()) {
+    case JSC::TypeUint8Clamped: return ImageDataStorageFormat::Uint8;
+    case JSC::TypeFloat16: return ImageDataStorageFormat::Float16;
+    default: RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+    }
+}
+
+template <typename From, typename To>
+struct TypedArrayItemConverter;
+
+template <>
+struct TypedArrayItemConverter<JSC::Uint8ClampedArray, JSC::Float16Array> {
+    static constexpr JSC::Float16Adaptor::Type convert(JSC::Uint8ClampedAdaptor::Type value)
+    {
+        return double(value) / 255.0;
+    }
+};
+
+template <>
+struct TypedArrayItemConverter<JSC::Float16Array, JSC::Uint8ClampedArray> {
+    static constexpr JSC::Uint8ClampedAdaptor::Type convert(JSC::Float16Adaptor::Type value)
+    {
+        auto d = double(value);
+        if (d <= 0)
+            return 0;
+        if (d >= 1)
+            return 255;
+        return d * 255.0;
+    }
+};
+
+template <typename ToTypeArray>
+static Ref<ToTypeArray> convertToTypedArray(const ImageDataArray& fromImageDataArray)
+{
+    return visitArrayBufferView(fromImageDataArray.arrayBufferView(), []<typename FromTypedArray>(FromTypedArray& fromTypedArray) {
+        if constexpr (std::is_same_v<FromTypedArray, ToTypeArray>)
+            return Ref(fromTypedArray);
+        else {
+            using Converter = TypedArrayItemConverter<FromTypedArray, ToTypeArray>;
+            auto length = fromTypedArray.length();
+            Ref<ToTypeArray> toArray = ToTypeArray::create(length);
+            for (size_t i = 0; i < length; ++i)
+                toArray->set(i, Converter::convert(fromTypedArray.item(i)));
+            return toArray;
+        }
+    });
+}
+
+Ref<JSC::Uint8ClampedArray> ImageDataArray::asUint8ClampedArray() const
+{
+    return convertToTypedArray<JSC::Uint8ClampedArray>(*this);
+}
+
+Ref<JSC::Float16Array> ImageDataArray::asFloat16Array() const
+{
+    return convertToTypedArray<JSC::Float16Array>(*this);
+}
```

Comment:
This interface is now ambiguous. The return value might be just a reference to the `ArrayBufferView` of the `ImageData` if no conversion happened. But it might be a new array converted from the existing `ArrayBufferView` if the requested format is not the same as the format of the `ImageData`. 

Is this covered by the specs? What should happen if JS modifies the converted array? Should this affect the stored array?

For example you create a Float16 ImageData and you requested a Uint8 array. Then you modified this Uint8 array. Should this modification be reflected in the stored ImageData Float16 array?

Reply:
From the JS API perspective, the user would use the `data` property or `ImageData`, which directly provides the `(Uint8ClampedArray or Float16Array)` object, no issue there.

`asUint8ClampedArray` effectively replaces the old **internal** `ImageData::data()` that used to always return a `Ref<Uint8ClampedArray>`, so at least the uint8-everywhere version (the default with the feature OFF) should work just like before.
And I believe these should go away once float16 is fully supported.

But for now, as far as I can see, all its uses are as a source of read-only pixels (for blobs, dataURL, put's...).
And in some of these cases, we'd want to avoid copies when not needed, and it would probably have a noticeable perf impact.

Happy to discuss further... Better naming maybe?

Reply:
You are right. I think I was confused. I thought both of the formats will be exposed to JS. But I think the created format will be exposed only to JS. The asUint8ClampedArray() will be used by `putImageData()` for now. right?